lib/blockchain.js

```diff
@@ -0,0 +1,189 @@
+const events = require('events');
+const {EventEmitter} = events;
+const dns = require('dns');
+const https = require('https');
+const fetch = require('node-fetch');
+const Web3 = require('web3');
+const {timeout} = require('../lib/util.js');
+
+const infuraProjectId = process.env.infuraProjectId;
+const polygonVigilKey = process.env.polygonVigilKey;
+
+const ethereumHost = 'ethereum.exokit.org';
+
+let addresses,
+  abis,
+  web3,
+  web3sockets,
+  contracts,
+  gethNodeUrl,
+  gethNodeWSUrl;
+
+const BlockchainNetworks = [
+  "mainnet",
+  "mainnetsidechain",
+  //"testnet",
+  "polygon",
+  //"testnetpolygon",
+];
+
+const loadPromise = (async() => {
+  const [
+    newAddresses,
+    newAbis,
+    ethereumHostAddress,
+    newPorts,
+  ] = await Promise.all([
+    fetch('https://contracts.webaverse.com/config/addresses.js').then(res => res.text()).then(s => JSON.parse(s.replace(/^\s*export\s*default\s*/, ''))),
+    fetch('https://contracts.webaverse.com/config/abi.js').then(res => res.text()).then(s => JSON.parse(s.replace(/^\s*export\s*default\s*/, ''))),
+    new Promise((accept, reject) => {
+      dns.resolve4(ethereumHost, (err, addresses) => {
+        if (!err) {
+          if (addresses.length > 0) {
+            accept(addresses[0]);
+          } else {
+            reject(new Error('no addresses resolved for ' + ethereumHost));
+          }
+        } else {
+          reject(err);
+        }
+      });
+    }),
+    (async () => {
+      return await new Promise((accept, reject) => {
+        const proxyReq = https.request('https://contracts.webaverse.com/config/ports.js', proxyRes => {
+          const bs = [];
+          proxyRes.on('data', b => {
+            bs.push(b);
+          });
+          proxyRes.on('end', () => {
+            accept(JSON.parse(Buffer.concat(bs).toString('utf8').slice('export default'.length)));
+          });
+          proxyRes.on('error', err => {
+            reject(err);
+          });
+        });
+        proxyReq.end();
+      });
+    })(),
+  ]);
+  addresses = newAddresses;
+  abis = newAbis;
+
+  const ports = newPorts;
+  gethNodeUrl = `http://${ethereumHostAddress}`;
+  gethNodeWSUrl = `ws://${ethereumHostAddress}`;
+
+  web3 = {
+    mainnet: new Web3(new Web3.providers.HttpProvider(
+      `https://mainnet.infura.io/v3/${infuraProjectId}`
+    )),
+    mainnetsidechain: new Web3(new Web3.providers.HttpProvider(
+      `${gethNodeUrl}:${ports.mainnetsidechain}`
+    )),
+    polygon: new Web3(new Web3.providers.HttpProvider(
+      `https://rpc-mainnet.maticvigil.com/v1/${polygonVigilKey}`
```

Comment:
We gotta be careful about this. Polygon is changing the endpoints on us.

Perhaps they should be abstracted, just like `abi.js`.

Reply:
Agreed. Should be abstracted to one source of truth.